src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -1732,7 +1732,7 @@ protected void CheckProjectDoesNotHaveProjectReference(ProjectId fromProjectId,
         /// </summary>
         protected void CheckProjectDoesNotHaveTransitiveProjectReference(ProjectId fromProjectId, ProjectId toProjectId)
         {
-            var transitiveReferences = GetTransitiveProjectReferences(toProjectId);
+            var transitiveReferences = this.CurrentSolution.GetProjectDependencyGraph().GetProjectsThatThisProjectTransitivelyDependsOn(toProjectId);
```

Comment:
Heads up on this change: @Pilchie's guess was correct: this was written prior to the project dependency graph existed. On one hand this is a "well, duh, of course do that!" fix, but the interesting effect would mean we are now going to have a lot more asking for transitive project references that previously wasn't. This, combined with the incremental updating above, means we might be caching this data when we don't exactly need it much anymore. In other words, this might hurt steady state memory more than it helps CPU. It'll reduce gen0 churn, but maybe that's cheaper than this?